### PR TITLE
Add initial support for IntrinsicFunction in Julia

### PR DESCRIPTION
**Pull request by Thirumalai-Shaktivel**

### CODE REVIEW

File: tests/reference/wat-abs_03-5d62cca.stdout [hunk 1]
```diff
@@ -11,30 +11,6 @@
...
-    (global $19 (mut i32) (i32.const 4))
-    (global $20 (mut i32) (i32.const 8))
-    (global $21 (mut i32) (i32.const 1))
-    (global $22 (mut i32) (i32.const 0))
-    (global $23 (mut i32) (i32.const 0))
-    (global $24 (mut i32) (i32.const 6))
-    (global $25 (mut i32) (i32.const -1))
-    (global $26 (mut i32) (i32.const 4))
-    (global $27 (mut i32) (i32.const 8))
-    (global $28 (mut i32) (i32.const 0))
```
Comment (Shaikh-Ubaid) [inline tests/reference/wat-abs_03-5d62cca.stdout:37]: @Thirumalai-Shaktivel do you know why global variables are being removed?

Comment (Thirumalai-Shaktivel) [reply tests/reference/wat-abs_03-5d62cca.stdout:37]: By any chance, is it possible to know what these global variables represent?

I thought it is something related to this change in Julia: https://github.com/lfortran/lfortran/pull/1500/files#diff-30a2e77476e8eaaeed0a2281d4fa431fa616dd3075edcf6b6623d396aac8b038L157

Comment (Shaikh-Ubaid) [reply tests/reference/wat-abs_03-5d62cca.stdout:37]: The shared link https://github.com/lfortran/lfortran/pull/1500/files#diff-30a2e77476e8eaaeed0a2281d4fa431fa616dd3075edcf6b6623d396aac8b038L157 seems to just open the difference tab of this PR. It does seem to point to any specific line of code for me. Can you trying sharing via some other way? Probably another link or screenshot?

Comment (Thirumalai-Shaktivel) [reply tests/reference/wat-abs_03-5d62cca.stdout:37]: Check the changes of this file: `tests/reference/julia-arrays_02-849d7b0.stdout`

Comment (Shaikh-Ubaid) [reply tests/reference/wat-abs_03-5d62cca.stdout:37]: I see. So, why are those functions being removed in `tests/reference/julia-arrays_02-849d7b0.stdout`?

It seems with the intrinsic function pass, intrinsic modules are being removed (I guess), right?

Comment (Thirumalai-Shaktivel) [reply tests/reference/wat-abs_03-5d62cca.stdout:37]: Yup, previously I think we were using the `function` definition from the `lfortran_intrinsics_math` module. Now, we create all the functions ourselves using the `intrinsic_function` pass.

Comment (Shaikh-Ubaid) [reply tests/reference/wat-abs_03-5d62cca.stdout:37]: >Yup, previously I think we were using the function definition from the lfortran_intrinsics_math module. Now, we create all the functions ourselves using the intrinsic_function pass.

Great! Thank you for the clarification. So, only function definitions from intrinsic modules are removed (with the module and its global variables still remaining in the ASR) or the whole intrinsic module is removed?

Comment (Thirumalai-Shaktivel) [reply tests/reference/wat-abs_03-5d62cca.stdout:37]: See https://github.com/lfortran/lfortran/pull/1480/files

**Review by czgdp1807 — CHANGES_REQUESTED** [review]
This makes sense to me but why can't we use the pass_manager APIs here as well? Can you try out using them instead of manually calling the pass.

File: src/libasr/codegen/asr_to_julia.cpp [hunk 2]
```diff
@@ -1885,9 +1886,21 @@ class ASRToJuliaVisitor : public ASR::BaseVisitor<ASRToJuliaVisitor>
...
+
+    LCompilers::PassOptions pass_options;
+    pass_options.runtime_library_dir = co.runtime_library_dir;
+    pass_options.mod_files_dir = co.mod_files_dir;
+    pass_options.include_dirs = co.include_dirs;
+    pass_options.always_run = false;
+    pass_options.verbose = co.verbose;
+    pass_manager.rtlib = co.rtlib;
+    pass_manager.apply_passes(al, &asr, pass_options, diag);
+
```
Comment (Thirumalai-Shaktivel) [inline src/libasr/codegen/asr_to_julia.cpp:1903]: This is the way to do it, right?

File: tests/reference/julia-arrays_05-f639a33.stdout [hunk 3]
```diff
@@ -1,10 +1,45 @@
 function main()
+    local __1_k::Int32 = 0
     local i::Int32 = 0
     local numbers::Array{Float32, 1} = [  1.50000000000000000e+00,   3.20000000000000018e+00,   4.50000000000000000e+00,   9.00000000000000022e-01,   7.20000000000000018e+00]
-    for i ∈ 1:5
+    __1_k = 1
```
Comment (Thirumalai-Shaktivel) [inline tests/reference/julia-arrays_05-f639a33.stdout:5]: There seems to be some bug in the ASRPass. Array: `numbers` is initialised three times.

Comment (czgdp1807) [reply tests/reference/julia-arrays_05-f639a33.stdout:5]: This seems to me like the output of loop_unroll.cpp pass. Can you apply that pass on the input code and see the ASR output of it? Use, `--pass=loop_unroll --show-asr`.

Comment (Thirumalai-Shaktivel) [reply tests/reference/julia-arrays_05-f639a33.stdout:5]: I was able to debug the issue here:
```Clojure
(Variable
    2
    numbers
    []
    Local
    (ArrayConstant
        [(RealConstant
            1.500000
            (Real 4 [])
        )
        ...
        ]
        (Real 4 [((IntegerConstant 1 (Integer 4 []))
        (IntegerConstant 5 (Integer 4 [])))])
        ColMajor
    )
    (ArrayConstant
        [(RealConstant
            1.500000
            (Real 4 [])
        )
        ...
        ]
        (Real 4 [((IntegerConstant 1 (Integer 4 []))
        (IntegerConstant 5 (Integer 4 [])))])
        ColMajor
    )
    Save
    (Real 4 [((IntegerConstant 1 (Integer 4 []))
    (IntegerConstant 5 (Integer 4 [])))])
    Source
    Public
    Required
    .false.
)
```
Here the `x.symbolic` and `x.value` are visited as `ArrayConstant` and in turn the values are initialised twice

Comment (Thirumalai-Shaktivel) [reply tests/reference/julia-arrays_05-f639a33.stdout:5]: I thought about this and I think we should set `x.m_value` to `null` if `x.m_symbolic_value` and `x.m_value` are same (ArrayConstant).
@czgdp1807 What do you suggest on this?

File: src/libasr/pass/init_expr.cpp [hunk 4]
```diff
@@ -140,6 +140,10 @@ class InitExprVisitor : public ASR::CallReplacerOnExpressionsVisitor<InitExprVis
             }
 
             ASR::Variable_t& xx = const_cast<ASR::Variable_t&>(x);
+            if (x.m_symbolic_value == x.m_value &&
+                    ASR::is_a<ASR::ArrayConstant_t>(*x.m_value)) {
+                xx.m_value = nullptr;
+            }
```
Comment (Thirumalai-Shaktivel) [inline src/libasr/pass/init_expr.cpp:146]: https://github.com/lfortran/lfortran/pull/1500#discussion_r1161097526

File: src/libasr/codegen/asr_to_julia.cpp [hunk 5]
```diff
@@ -1812,6 +1815,16 @@ class ASRToJuliaVisitor : public ASR::BaseVisitor<ASRToJuliaVisitor>
...
 
+    void visit_ArrayBound(const ASR::ArrayBound_t& x) {
+        visit_expr(*x.m_v);
+        std::string var_name = src;
+        if (x.m_bound == ASR::arrayboundType::LBound) {
+            src = "1";
+        } else if( x.m_bound == ASR::arrayboundType::UBound ) {
+            src = "size(" + var_name +", 1)";
+        }
+    }
```
Comment (Thirumalai-Shaktivel) [inline src/libasr/codegen/asr_to_julia.cpp:1826]: I'm not familiar with Julia, is this correct way to ArrayBound?

File: tests/reference/julia-array3-ab6fef3.stdout [hunk 6]
```diff
@@ -1,8 +1,21 @@
...
+    __1_k = __1_k + 1
+    a[__1_k] = Float32(3)
+    __1_k = __1_k + 1
+    __1_k = 1
+    b[__1_k] = Float32(1)
+    __1_k = __1_k + 1
+    b[__1_k] = Float32(2)
+    __1_k = __1_k + 1
+    b[__1_k] = Float32(3)
+    __1_k = __1_k + 1
```
Comment (certik) [inline tests/reference/julia-array3-ab6fef3.stdout:18]: I think we should not be lowering high level constructs, but rather keep constructs high level and use an equivalent high level Julia feature.

Comment (Thirumalai-Shaktivel) [reply tests/reference/julia-array3-ab6fef3.stdout]: Done!

File: src/libasr/pass/init_expr.cpp (hunk 4 above)
Comment (czgdp1807) [reply src/libasr/pass/init_expr.cpp:146]: If you are not using this pass in Julia then please don't do this change. If you want to keep it then also test it with LPython by making a PR there.

**Review by czgdp1807 — CHANGES_REQUESTED** [review]

File: tests/reference/julia-arrays_05-f639a33.stdout (hunk 3 above)
Comment (czgdp1807) [reply tests/reference/julia-arrays_05-f639a33.stdout]: Depends. If the variable is a constant then we should keep `x.m_value` and backends like LLVM, x86 should use that value directly. However backends like, Julia, C shouldn't be using `x.m_value` if `--fast` mode is not enabled. If `--fast` mode is enabled then all backends should use `x.m_value` whenever its available. So the question is not, which ones should be kept `nullptr` but the question is which one to use and in which backend and under which context. So unless extremely necessary we shouldn't be setting `x.m_symbolic_value` and `x.m_value` to `nullptr` in `init_expr.cpp`.

**Comment by Thirumalai-Shaktivel** [comment]
Cool, I created an issue for it, https://github.com/lcompilers/lpython/issues/1687

File: tests/reference/julia-arrays_01-81f01a2.stdout [hunk 7]
```diff
@@ -2,7 +2,9 @@ function main()
     local a = Array{Int32, 1}(undef, 3)
     local b = Array{Int32, 1}(undef, 4)
     local i::Int32 = 0
-    for i ∈ 1:3
+    i = 1 - 1
+    while i + 1 ≤ 3
+        i = i + 1
```
Comment (certik) [inline tests/reference/julia-arrays_01-81f01a2.stdout:7]: This removes a for loop and replaces it with a while loop. We should not do that, let's keep the for loop.

File: tests/reference/julia-case_01-cbdd0aa.stdout [hunk 8]
```diff
@@ -5,15 +5,21 @@ function main()
...
+            if i == 3
+                out = 30
+                println("3")
+            else
+                if i == 4
+                    out = 40
+                    println("4")
+                end
+            end
+        end
```
Comment (certik) [inline tests/reference/julia-case_01-cbdd0aa.stdout:22]: The if structure seems broken, I think there was some logic before to generate `elseif` which is more readable.

Comment (certik) [reply tests/reference/julia-case_01-cbdd0aa.stdout:22]: I think it's the "case" switch that was lost and rewritten to if statements in ASR.

**Comment by certik** [comment]
@Thirumalai-Shaktivel what bug are we trying to fix?

We need to remove the "loop" and "switch" ASR passes, those should not be applied. 

Let's do it the other way round: let's not apply any passes and then let's justify every pass that should be applied.

**Comment by Thirumalai-Shaktivel** [comment]
I was trying to apply ASR Passes for Julia Backend.

**Comment by Thirumalai-Shaktivel** [comment]
This is ready, kindly review it.

File: src/libasr/codegen/asr_to_julia.cpp [hunk 9]
```diff
@@ -1880,6 +1881,8 @@ Result<std::string>
 asr_to_julia(Allocator& al, ASR::TranslationUnit_t& asr, diag::Diagnostics& diag)
 {
     ASRToJuliaVisitor v(al, diag);
+    LCompilers::PassOptions pass_options;
+    pass_replace_intrinsic_function(al, asr, pass_options);
```
Comment (czgdp1807) [inline src/libasr/codegen/asr_to_julia.cpp:1885]: There is one more way to do it. Create a vector `std::vector<std::string> julia_passes = {"intrinsic_functions"}`. Make the following method public (i.e., remove the `_` and shift it to public scope),

https://github.com/lfortran/lfortran/blob/523c4723feba77241d4de241afc1fac95ba3bfe6/src/libasr/pass/pass_manager.h#L100

Then call, `pass_manager.apply_passes(al, asr, julia_passes, pass_options, diag)`

What I am recommending is to not call passes manually but instead use pass_manager for your use case and feel free to modify pass_manager to suit your needs. Everything related to passes should be present in pass_manager and we should be using it to call any ASR pass.

**Review by czgdp1807 — CHANGES_REQUESTED** [review]

**Comment by certik** [comment]
> I was trying to apply ASR Passes for Julia Backend.

I think rather we are ensuring the intrinsic function call gets called?

But shouldn't C++ and Julia backends not call this call and rather use Julia "sin" function directly?

File: src/libasr/codegen/asr_to_julia.cpp (hunk 9 above)
Comment (Thirumalai-Shaktivel) [reply src/libasr/codegen/asr_to_julia.cpp:1885]: Cool, this makes sense. I will make the required changes.
Thank you!

**Comment by czgdp1807** [comment]
> But shouldn't C++ and Julia backends not call this call and rather use Julia "sin" function directly?

That makes most sense. In fact, that's the purpose of `IntrinsicFunction`? To leave it to the backend on how they want to implement and call an intrinsic, either by using their own intrinsics (like, C++ has `sin`, Julia has its own `sin` or like LLVM which will call into our C implementation of sin present in `lfortran_intrinsics.c`). So, `intrinsic_functions.cpp` should be applied only to those backends which don't have their own implementation for a given intrinsic.

**Comment by Thirumalai-Shaktivel** [comment]
Exactly, why didn't I think about this?
It reduces the overhead to compute the value, instead, we should make use of their built-in function itself.
I'm +1 on this. I will try to implement this

**Comment by certik** [comment]
Exactly.

I think what we will also need to call the intrinsic function pass for some functions but not all. In particular, we will have hundreds or even thousands of intrinsics, various special functions etc. For the vast majority of them you want to call the pass, but for the elementary like trig functions you don't.

So we probably want to enable calling this pass, but selecting which functions to transform (or perhaps "not transform"), those would we handle in the backend ourselves.

File: src/libasr/codegen/asr_to_julia.cpp [hunk 10]
```diff
@@ -1874,6 +1875,19 @@ class ASRToJuliaVisitor : public ASR::BaseVisitor<ASRToJuliaVisitor>
         std::string out = indent + "// FIXME: File Read\n";
         src = out;
     }
+
+    void visit_IntrinsicFunction(const ASR::IntrinsicFunction_t &x) {
+        if(x.m_value) {
+            visit_expr(*x.m_value);
+            return;
+        }
```
Comment (certik) [inline src/libasr/codegen/asr_to_julia.cpp:1883]: I don't think we want to do this, unless in `--fast` mode.

File: src/libasr/codegen/asr_to_julia.cpp [hunk 11]
```diff
@@ -1874,6 +1875,19 @@ class ASRToJuliaVisitor : public ASR::BaseVisitor<ASRToJuliaVisitor>
...
+        std::string out;
+        out = to_lower(ASRUtils::get_intrinsic_name(x.m_intrinsic_id));
+        LCOMPILERS_ASSERT(x.n_args == 1);
+        visit_expr(*x.m_args[0]);
+        out += "(" + src + ")";
+        src = out;
```
Comment (certik) [inline src/libasr/codegen/asr_to_julia.cpp:1889]: In here I would do it explicitly based on the given function, for example only for `abs`, `sin`, `cos`, etc. Julia might have different names for some of the intrinsic functions.

**Review by certik — APPROVED** [review]
I think that this is good.

**Comment by Thirumalai-Shaktivel** [comment]
Thanks for the review!
Let's merge this PR. I will submit the IntrinisicFunction changes to LPython as well.

**Pull request merged**